task4/src/main/scala/com/bootcamp/task/Homework.scala

```diff
@@ -0,0 +1,129 @@
+package com.bootcamp.task
+
+import scala.annotation.tailrec
+
+object Homework {
+
+	// try to implement min different ways (fold, reduce, recursion)
+	def min(list: List[Int]): Option[Int] = {
+		@tailrec
+		def recurrentMin(min: Int, tail: List[Int]): Int = tail match {
+			case Nil => min
+			case head :: tail => recurrentMin(math.min(head, min), tail)
+		}
+		list match {
+			case Nil => None
+			case current :: tail => Some(recurrentMin(current, tail))
+		}
+	}
+
+	// Implement scanLeft (not using scans ofc)
+	def scanLeft[T](zero: T)(list: List[T])(f: (T, T) => T): List[T] = {
+		list match {
+			case Nil => List(zero)
+			case l =>
+				val (_, history) = l.foldLeft((zero, List(zero))) { case ((acc, history), item) =>
+					val res = f(acc, item)
+					(res, history :+ res)
+				}
+				history
+		}
+	}
+
+	// https://twitter.com/allenholub/status/1357115515672555520/photo/1
+	// pass the interview
+	def count(s: String): List[(Char, Int)] = {
+		@tailrec
+		def countRec(current: Char, count: Int, str: List[Char], res: List[(Char, Int)]): List[(Char, Int)] = {
+			str match {
+				case Nil => res :+ (current, count)
+				case char :: tail =>
+					val (nextCount, nextRes) = if(current == char) (count + 1, res) else (1, res :+ (current, count))
+					countRec(char, nextCount, tail, nextRes)
+			}
+		}
+		s.toList match {
+			case Nil => Nil
+			case c :: tail => countRec(c, 1, tail, Nil)
+		}
+	}
+
+
+	// hometask:
+	// https://leetcode.com/problems/running-sum-of-1d-array/
+	def runningSum(nums: Array[Int]): Array[Int] = {
+		if (nums == null || nums.length == 0) nums
```

Comment:
in scala we dont check for nulls, scala people dont usually pass nulls so usually it is not a problem

Reply:
Yeah, this was first leetcode task, wasn't sure whether they pass a null array or not

Reply:
BTW, I guess there's not "Nil" value for arrays, to show it;s an empty one?

Reply:
.isEmpty is fine i think and it works for everything

Reply:
Thanks!